Rover hold/loiter

ealdaz-seesai opened the pull request:
This is my first pull request, so unsure as to whether I should  raise an Issue before creating a pull request, but I'm hoping that this explanation will be sufficiently clear, any feedback welcome. 

**Describe problem solved by this pull request**
When using HOLD/LOITER the rover would not hold position but instead circle/oscillate around the hold position. This was caused by the fact that the current algorithm sets velocity to zero when distance to waypoint is under a certain threshold, the vehicle would decelerate, but not enough to stop before it once again exceeded the threshold. At this point it would then try to return to the hold position and the process would repeat. 

**Describe your solution**
In the proposed solution once the distance to the waypoint is under a threshold the velocity is set to zero until a new position waypoint is requested that is more than the threshold distance away from the previous waypoint. This way the vehicle always stops.

**Test data / coverage**
This was tested with both px4_sitl and a pixhawk4 rover running the  Aion RObotics R1 UGV (Tank type/Differential drive)

ealdaz-seesai commented:
This is an example of what happens when selecting HOLD in QGroundControl
![image](https://user-images.githubusercontent.com/35691522/71261369-1ae9f300-2335-11ea-927a-50c83faccd53.png)

ealdaz-seesai commented:
In the proposed implementation the rover stops when HOLD is requested:
![image](https://user-images.githubusercontent.com/35691522/71261700-f2aec400-2335-11ea-8fd1-9d827644aac9.png)

julianoes commented:
@ItsTimmy would be great if you could review this one, thanks!

ItsTimmy commented on src/modules/rover_pos_control/RoverPositionControl.cpp:
```diff
@@ -513,12 +518,13 @@ RoverPositionControl::run()
+
+			} else {
+
+				_actuator_controls_pub = orb_advertise(ORB_ID(actuator_controls_0), &_act_controls);
```
Why was this changed? Shouldn't we still only publish actuator controls when in the right mode?

ealdaz-seesai replied:
As I understand it RoverPositionControl is only called when we are actively in position mode,  see line 422

```
if (!manual_mode && _control_mode.flag_control_position_enabled) {
if (control_position(current_position, ground_speed, _pos_sp_triplet)) {
```

So didn't really understand why the flags for velocity or attitude control are checked inside the position loop, perhaps it's inherited from an older implementation and hasn't been cleaned before?

Jaeyoung-Lim replied:
@ealdaz-seesai This is not true, although the name of the file is rover position control, it can be used in different modes such as offboard modes. Those functions are used to support different setpoints(position, velocity, attitude, actuator) in offboard mode.

julianoes commented:
@ealdaz-seesai it would be great if you could follow up on the things @ItsTimmy mentioned and make sure the CI tests pass.

ealdaz-seesai commented:
Will do Julian, currently travelling with very bad coverage. Back Monday.
Thanks for pushing this along.

Eduardo

---

Eduardo Aldaz Carroll
CTO
sees.ai

E: ealdaz@sees.ai
M: +44 (0) 7543 519191
Skype: e.aldaz-carroll

> On 20 Feb 2020, at 15:13, Julian Oes <notifications@github.com> wrote:
> 
> @ealdaz-seesai <https://github.com/ealdaz-seesai> it would be great if you could follow up on the things @ItsTimmy <https://github.com/ItsTimmy> mentioned and make sure the CI tests pass.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/PX4/Firmware/pull/13767?email_source=notifications&email_token=AIQJYAUVLQYV7R6GILEX7KLRD2F2BA5CNFSM4J56HLF2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMOHGXY#issuecomment-589067103>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AIQJYAQV3SY2SFFPRYCODM3RD2F2BANCNFSM4J56HLFQ>.
>

ealdaz-seesai commented on src/modules/rover_pos_control/RoverPositionControl.hpp:
```diff
@@ -99,10 +98,10 @@ class RoverPositionControl : public ModuleBase<RoverPositionControl>, public Mod
-	uORB::Publication<position_controller_status_s>	_pos_ctrl_status_pub{ORB_ID(position_controller_status)};
-	uORB::Publication<actuator_controls_s>		_actuator_controls_pub{ORB_ID(actuator_controls_0)};
+	orb_advert_t	_pos_ctrl_status_pub{nullptr};		/**< navigation capabilities publication */
+	orb_advert_t    _actuator_controls_pub{nullptr};	/**< actuator controls publication */
```
To be honest I found examples of both styles in the repository, so from memory I used whatever was been used for Rover. Can you modify them to be correct, or is that something I need to do on my side?

ealdaz-seesai replied:
Changed to C++ API

ealdaz-seesai commented on src/modules/rover_pos_control/RoverPositionControl.cpp:
```diff
@@ -462,7 +462,12 @@ RoverPositionControl::run()
+
+					} else {
+						_pos_ctrl_status_pub = orb_advertise(ORB_ID(position_controller_status), &pos_ctrl_status);
+					}
```
Done

Jaeyoung-Lim commented:
@ealdaz-seesai Could you resolve the conflicts of this PR?

ealdaz-seesai commented:
Of course, hadn't realized there were conflicts, might be due to some new changes to the file. Let me check.

ealdaz-seesai commented:
All tests passing except some MacOS, complaining about implicit conversion from float to double in calls to function get_distance_to_next_waypoint  using Vector2f variables

ealdaz-seesai commented:
Is there anything else required to merge this?

ealdaz-seesai commented:
bump 
@ItsTimmy @dagar @julianoes @Jaeyoung-Lim anything else necessary to merge this?
It seems to me it's ready to go?
Thanks!

ealdaz-seesai commented:
Hello Julian
I merged 3 weeks ago as some of the files that I had originally modified 4 months ago had been changed since. All tests passed at the time, and couldn't see any conflicts.
Regardless I've merged again without any conflicts.
All checks passed except for the `Jenkins Compile MacOS` one. 
I don't understand what the error is, doesn't seem to be related to my code for what I can tell.
Could you check please?
Thx

ealdaz-seesai commented:
Also not sure I understand the submodule point you were making, I haven't changed any submodule.

julianoes commented:
I'm really confused about the state of this PR, I can't get it merged properly either. It's odd.

Some submodules are wrong, see here:
![Selection_048](https://user-images.githubusercontent.com/1419688/81044465-b038ee80-8eb4-11ea-96c6-6a2d5200c86e.png)

Let me try to apply the diff and create one new clean commit.

julianoes commented:
I have the diff applied cleanly but I can't push it to your fork.

Could you make the checkmark as described here?
https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

ealdaz-seesai commented:
I’m really sorry Julian, but I can’t find that check box anywhere on the Pull request page….

Any tips on how to find it, it’s like it’s not been enabled ...




Eduardo

---

Eduardo Aldaz Carroll
CTO
sees.ai

E: ealdaz@sees.ai
M: +44 (0) 7543 519191
Skype: e.aldaz-carroll

> On 5 May 2020, at 08:48, Julian Oes <notifications@github.com> wrote:
> 
> 
> I have the diff applied cleanly but I can't push it to your fork.
> 
> Could you make the checkmark as described here?
> https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork <https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/PX4/Firmware/pull/13767#issuecomment-623908067>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AIQJYAVB5ZGN66MXFEXQEFLRP7ADVANCNFSM4J56HLFQ>.
>

julianoes commented:
> Is it because the repository is not my personal one but part of an organisation?

That's probably why. Do you want to create a new PR with the clean diff or should I do it?

ealdaz-seesai commented:
I’ll give it a go

Eduardo

---

Eduardo Aldaz Carroll
CTO
sees.ai

E: ealdaz@sees.ai
M: +44 (0) 7543 519191
Skype: e.aldaz-carroll

> On 6 May 2020, at 08:26, Julian Oes <notifications@github.com> wrote:
> 
> 
> Is it because the repository is not my personal one but part of an organisation?
> 
> That's probably why. Do you want to create a new PR with the clean diff or should I do it?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/PX4/Firmware/pull/13767#issuecomment-624486571>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AIQJYAQVHUB6PIRTKDDUDOTRQEGLLANCNFSM4J56HLFQ>.
>

ealdaz-seesai commented:
OK, I've figured out what the issue was. Those submodules were added to my git branch 3 weeks ago and I didn't realize. I've removed them now. I'm hoping it should all be OK now.

julianoes commented:
Great, thanks @ealdaz-seesai.

@Jaeyoung-Lim could you please give this a review?

Jaeyoung-Lim commented on src/modules/rover_pos_control/RoverPositionControl.hpp:
```diff
@@ -113,15 +114,15 @@ class RoverPositionControl : public ModuleBase<RoverPositionControl>, public Mod
-	vehicle_attitude_s			_vehicle_att{};
-	sensor_combined_s			_sensor_combined{};
+	actuator_controls_s			    _act_controls{};		/**< direct control of actuators */
+	vehicle_attitude_s              _vehicle_att{};
```
Please fix the indentation errors

ealdaz-seesai replied:
OK, will do. 
I run the fix_code_style.sh, thought that took care of that sort of thing. But just realized that CLion was using tabs as spaces, will sort it out.

Jaeyoung-Lim replied:
Thanks!

Jaeyoung-Lim commented on src/modules/rover_pos_control/RoverPositionControl.hpp:
```diff
@@ -113,15 +114,15 @@ class RoverPositionControl : public ModuleBase<RoverPositionControl>, public Mod
 	uORB::Subscription	_parameter_update_sub{ORB_ID(parameter_update)};
 
-	manual_control_setpoint_s		_manual{};			/**< r/c channel data */
+	manual_control_setpoint_s		_manual{};			    /**< r/c channel data */
```
unnecessary indentation changes

Jaeyoung-Lim replied:
White spaces are still not addressed

Jaeyoung-Lim commented on src/modules/rover_pos_control/RoverPositionControl.hpp:
```diff
@@ -113,15 +114,15 @@ class RoverPositionControl : public ModuleBase<RoverPositionControl>, public Mod
-	sensor_combined_s			_sensor_combined{};
+	actuator_controls_s			    _act_controls{};		/**< direct control of actuators */
+	vehicle_attitude_s              _vehicle_att{};
+	sensor_combined_s				_sensor_combined{};
```
same here

Jaeyoung-Lim replied:
White spaces are still not addressed

Jaeyoung-Lim commented on src/modules/rover_pos_control/RoverPositionControl.cpp:
```diff
@@ -406,7 +407,7 @@ RoverPositionControl::run()
 
 		/* only run controller if position changed */
-		if (fds[0].revents & POLLIN) {
+		if (fds[0].revents & POLLIN || fds[4].revents & POLLIN) {
```
I think we can use just local_position (`fds[4]`) and remove `fds[0]`

ealdaz-seesai replied:
Ha, pushing my very average git skills with this...never tried squashign commits before, I'll google and give it a try...

ealdaz-seesai replied:
Is local position also available if only using GPS?

ealdaz-seesai commented:
Hi @Jaeyoung-Lim , Ok so I've really tried, I've rebased and squashed, but still all the commits appear. I tried rebasing again and then all sorts of conflicts happen, I think it's got to do probably with the fact that I've merged twice already in the last months. I'm not comfortable enough with git to understand what is really going on, and I'm afraid I can't really spend anymore time with this.

ealdaz-seesai commented:
OK, so it was annoying me to leave it half done. So decided to give it one last go and rebase again and skip some commits and it has worked.
I've tested in SITL and all works as intended.

I did notice a new "feature" in this latest master and it's that SITL rover complains that no power is unavailable and doesn't let you arm. I had to disable power check flag. (CBRK_SUPPLY_CHECK).

Note that I couldn't reboot either from QGroundcontrol... i'm pretty sure I could before... 
So a couple of things to add as a future PR...
```
pxh> WARN  [PreFlightCheck] system power unavailable
````

ealdaz-seesai commented:
Jenkins SITL Tests fail, but not sure why... would any of you know ?

julianoes commented:
@Jaeyoung-Lim would be great if you could follow up.

Jaeyoung-Lim commented:
@julianoes Jenkins failure is not related to this PR. There are a few comments that are not addressed from @ealdaz-seesai Otherwise looks good to me

julianoes commented:
@Jaeyoung-Lim you're call if you want those addressed or merge anyway.

julianoes commented:
Thanks @ealdaz-seesai !

ealdaz-seesai commented:
Thank you both !  A pleasure!

Eduardo

---

Eduardo Aldaz Carroll
CTO
sees.ai

E: ealdaz@sees.ai
M: +44 (0) 7543 519191
Skype: e.aldaz-carroll

> On 13 May 2020, at 12:09, Julian Oes <notifications@github.com> wrote:
> 
> 
> Thanks @ealdaz-seesai <https://github.com/ealdaz-seesai> !
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/PX4/Firmware/pull/13767#issuecomment-627912810>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AIQJYAXPC6D2OWLWDDBHAYTRRJ5VXANCNFSM4J56HLFQ>.
>
